### PR TITLE
Clean deps from unloadable modules

### PR DESCRIPTION
**Pull request by graupe**
My attempt at solving #26.
 
 I have to admit, I only halve know what I'm doing here.
 
Please mind that I set the `CodeEvaluation.in_mix_compile?()`-Mock to return `false` in order to mimic the error case described in the linked issue.

I assume the actual fix should be for `in_mix_compile?` to be correctly return `true` during compilation phase?

### CODE REVIEW

**Comment by xpgdk** [comment]
I am experimenting with an alternative implementation of `CodeEvaluation.in_mix_compile?()`:
```
  def in_mix_compile? do
    :ets.info(Mix.State) != :undefined
  end
```

That seems to solve the problem, but I'm not sure if it is the correct way of doing it.

**Comment by IvanRublev** [comment]
Hi folks,
Thanks for rolling this forward. The library [can run interactively](https://github.com/IvanRublev/Domo/blob/f2c70e6e5443900e55430ce2f60c2fcf4b043456/lib/domo.ex#L112) in iex and livebook, so the idea of `in_mix_compile?` is to distinguish the case when it's for compilation. The [current implementation](https://github.com/IvanRublev/Domo/blob/f2c70e6e5443900e55430ce2f60c2fcf4b043456/lib/domo/code_evaluation.ex#L7) was confirmed by Jose Valim, so let's keep it like that for compatibility.

I'll have a detailed look tomorrow.

**Comment by IvanRublev** [comment]
Hey folks, 
The following might seem unprofessional. It's not that. 
After reading the code, I found that I was wrong. The deletion of unloadable dependencies is already there implicitly by error of type reading - [here is the test](https://github.com/IvanRublev/Domo/blob/master/test/domo/type_ensurer_factory/dependency_resolver_test.exs#L266). At the same time, the work of @graupe is valuable to name the things explicitly to find this quicker next time. So it'd be good to put this as a refactoring after finding the root cause of a bug. Now to the root cause.

How I understood the trace of the crash:
* Enter to `build_ensurer_modules/2`
* `remove_modules_with_unloadable_types/1` is expected to return list of the dependencies [skipping](https://github.com/IvanRublev/Domo/blob/master/lib/domo/type_ensurer_factory/dependency_resolver.ex#L130) modules which types can't be read.
* the crash happens after failing to read types in `beam_types/1` when unexpectedly `in_mix_compile?` returns `false` and [calls](https://github.com/IvanRublev/Domo/blob/master/lib/domo/type_ensurer_factory/module_inspector.ex#L60) to unloaded `ResolvePlanner` server.
* The server is not there because it runs only in iex, livebook interactive mode to hold the types information. Under the mix compile mode we rely on information from beam files, so the server intentionally stopped [here](https://github.com/IvanRublev/Domo/blob/master/lib/mix/tasks.compile.domo_compiler.ex#L64) before entering into `build_ensurer_modules/2`. 

To summarise, the `in_mix_compile?` calls Elixir's `Code.can_await_module_compilation?()`, which unexpectedly returns `false` in the case of mix compile. And it's trusted to return `true` in this case. The behaviour is interesting because this function [was recommended](https://groups.google.com/g/elixir-lang-core/c/vMJp5Jc5WiA/m/xc7s93X8CQAJ) to distinguish between interactive iex and livebook runs and mix compile runs. 

Let's step back and reproduce the bug. @xpgdk can you help with an example project reproducing the bug?

**Comment by xpgdk** [comment]
@IvanRublev I have created a very minimal project and a description to reproduce the bug: https://github.com/Qarma-inspect/domo_clean_deps
Let me know if there is anything else I can do.

**Comment by IvanRublev** [comment]
Hey @graupe  can you please have a look at the example project reproducing the bug posted above and find out what's going on with the `Code.can_await_module_compilation?()` behaviour?

**Comment by graupe** [comment]
Hi there.

I invested quite some time to understand the code base. Would need some explaining to fully grasp it.

What I found with my printf-style debugging is, that `CodeEvaluation.in_mix_compile?` is only `true` if actually called from within an `elixirc` compile run. The `domo_compiler` pass will not have this be set to `true`. (`Code.can_await_module_compilation?` will simply check, if the `:error_handler` of the running process is `Kernel.ErrorHandler`, indicating the presence of the `Kernel.PrallelCompiler`)

For lack of better ideas, I suggest reversing the logic in `ModuleInspector` from
```elixir
        if CodeEvaluation.in_mix_compile?() do
          {:error, {:no_beam_file, module}}
        else
          ResolvePlanner.get_types(:in_memory, module)
        end
```

to

```elixir
        if ResolvePlanner.started?(:in_memory) do
          ResolvePlanner.get_types(:in_memory, module)
        else
          {:error, {:no_beam_file, module}}
        end
```
with
```elixir
defmodule ResolvePlanner do
 ...
 def started?(plan_path), do: GenServer.whereis(via(plan_path)) != nil
 ...
```

To sneak a question in: How is it, that this code can expect the `:in_memory`-ResolvePlanner to be running? And if we can expect it to be running, shouldn't there rather be a way to pass `plan_path` somehow, to make it work the same in all instances, like so `ResolverPlanner.get_types(plan_path, module)`?

**Comment by IvanRublev** [comment]
Hi @graupe, thanks for focusing on the code!

I finally realised what the issue. Turns out that the `:error_handler` (that is for `Code.can_await_module_compilation?`) is set up only if the Elixir compiled any file. And in our case, Elixir first interprets the code -> finds nothing to compile (module was removed!) -> so flag is not set.

The `:in_memory` mode is for the IEx and LiveBook when all modules and appropriate `TypeEnsurers` are built from scratch during the interactive session in memory only. And the regular mode is for the case when we already have modules as BEAM files on the disk before the start of Domo. I think modules can be unloaded in both cases.

I agree with the reverse of the logic; And let's add two things:
* the comment that `CodeEvaluation.in_mix_compile?` wouldn't work because nothing is to be compiled in case of the module removal
* unit test the behaviour

**Comment by graupe** [comment]
Hi!

After some eventful days, I come back to this PR.

Work log is roughly as follows:

* removed commit 37c9b5d, as it wasn't useful (but actually gave rise to the question, if all the tests should be run for both cases `in_mix_compile?`: `true` and `false`)
* added the suggested logic reversal
* added comment describing issue with `CodeEvaluation.in_mix_compile?`
* added two dedicated tests to verify the behavior
* amended two tests to additionally `expect/refute` calls to `ResolvePlanner`
* added over the top check for `started?` function

Please let me know, what you think. My  knowledge of ExUnit and tooling is not advanced enough to do much more than this, unless guided, sorry.

**Comment by IvanRublev** [comment]
Hi @graupe,
Thanks for the code. It looks good to me and makes the example project compile successfully! I added an acceptance test with f4b9bc7 that was failing on master and now passed with your change.

And big thanks to @xpgdk for the example project. I'm now merging this to the master to do a release.
